src/rulesets/Base/indexing.jl

```diff
@@ -54,38 +54,77 @@ end
 
 
 #####
-##### getindex
+##### getindex(::AbstractArray)
 #####
 
 function frule((_, ẋ), ::typeof(getindex), x::AbstractArray, inds...)
     return x[inds...], ẋ[inds...]
 end
 
-function rrule(::typeof(getindex), x::Array{<:Number}, inds...)
+function rrule(::typeof(getindex), x::AbstractArray, inds...)
     # removes any logical indexing, CartesianIndex etc
     # leaving us just with a tuple of Int, Arrays of Int and Ranges of Int
     plain_inds = Base.to_indices(x, inds)
     y = getindex(x, plain_inds...)
     function getindex_pullback(ȳ)
-        function getindex_add!(Δ)
-            # this a optimizes away for simple cases
-            for (ȳ_ii, ii) in zip(ȳ, Iterators.product(plain_inds...))
-                Δ[ii...] += ȳ_ii
-            end
-            return Δ
-        end
-
-        x̄ = InplaceableThunk(
-            getindex_add!,
-            @thunk(getindex_add!(zero(x))),
+        xthunk = InplaceableThunk(
+            x̄ -> ∇getindex!(x̄, x, unthunk(ȳ), plain_inds...),
+            @thunk(∇getindex(x, unthunk(ȳ), plain_inds...)),
         )
-        īnds = broadcast(Returns(NoTangent()), inds)
-        return (NoTangent(), x̄, īnds...)
+        nots = map(Returns(NoTangent()), inds)
+        return (NoTangent(), xthunk, nots...)
     end
-
     return y, getindex_pullback
 end
 
+"""
+    ∇getindex(x, dy, inds...)
+
+For the `rrule` of `y = x[inds...]`, this function is roughly 
+`setindex(zero(x), dy, inds...)`, returning the array `dx`.
+Differentiable. Includes `ProjectTo(x)(dx)`.
+"""
+function ∇getindex(x::AbstractArray{<:Number}, dy, inds...)
+    # It's unfortunate to close over `x`, but `similar(typeof(x), axes(x))` doesn't 
+    # allow `eltype(dy)`, nor does it work for many structured matrices.
+    dx = fill!(similar(x, eltype(dy), axes(x)), 0)
+    ∇getindex!(dx, x, dy, inds...)
+    return ProjectTo(x)(dx)  # since we have x, may as well do this inside, not in rules
+end
+function ∇getindex(x::AbstractArray, dy, inds...)
+    # Since we have `x`, we can also handle arrays of arrays.
+    dx = map(zero, x)  # this ignores type of dy, TODO?
```

Comment:
I think we should pull all out zero creation out into a function called
`dzero(structure_x, element_y)`
So we would have:
```julia
dzero(x::AbstractArray{<:Number}, dy) = fill!(similar(x, eltype(dy), axes(x)), 0)
dzero(x::AbstractArray{<:AbstractArray}, dy<:AbstractArray) = map(xi, dyi->dzero(xi, dyi), x, dy)

dzero(x::Array{T}, dy::T) where T<:Number = zero(x)
# Maybe should be a 1 arg dzero that only takes element_dy ?
dzero(x::Number, y) = zero(y)
dzero(x, y) = ZeroTangent()


```


These get nightmarish fast...
I feel like if we had https://github.com/JuliaLang/julia/issues/38064 it would work better.
There is some code in that thread

Reply:
Maybe
```julia
function _setindex_zero(xs::AbstractArray, dy, inds::Integer...)
  dx = similar(xs, typeof(dy), axes(xs))
  return map!(zero, dx, dx)
end
function _setindex_zero(xs::AbstractArray, dy, inds...)
  dx = similar(xs, eltype(dy), axes(xs))
  return map!(zero, dx, dx)
end
```
but this gives 
```
ERROR: MethodError: Cannot `convert` an object of type ZeroTangent to an object of type Tangent{Base.Fix1, NamedTuple{(:x,), Tuple{Float64}}}
```

Reply:
Fixed. But I think it has to depend on the indices, to know whether to use `eltype(dy)` or `typeof(dy)`